Update Swift grammar

NinjaLikesCheez opened the pull request:
This change updates the grammar for Swift to use the latest version. In short, it adds semgrep support for:

* `#unavaliable` token
* if/switch expressions
* & reference operator
* parenthesized type expressions (such as `(some SomeType).foo` or `(any SomeType).bar`
* macro declarations
* `await` for loops
* `willSet/didSet` blocks
* `borrowing/consuming/distributed/package` keywords
* parameter packs

It also updates support for: 

* try operators postfix operator changes
* Equal sign changes

Please note, this is my **first** contribution to Semgrep and my first time writing OCaml - please please please review this very carefully! I know not what I am doing. 

Related to https://github.com/semgrep/ocaml-tree-sitter-semgrep/pull/523

NinjaLikesCheez committed:
Add #unavaliable token

NinjaLikesCheez committed:
Add support for if/switch expressions

[SE-0380](https://github.com/alex-pinkus/tree-sitter-swift/commit/f3063c6c25f71061421f456883eb0b55b56159c0)

NinjaLikesCheez committed:
Add & as an referencable operator

NinjaLikesCheez committed:
Map parenthesized type expressions for navigation expressions

NinjaLikesCheez committed:
Update comment for unary expressions

NinjaLikesCheez committed:
Fix switch expressions test

NinjaLikesCheez committed:
Add macro support

NinjaLikesCheez committed:
Add support for try operators

NinjaLikesCheez committed:
Add support for for_statement_collection

NinjaLikesCheez committed:
Add support for willSet/didSet blocks

NinjaLikesCheez committed:
Add support for Equal_sign

NinjaLikesCheez committed:
Add support for optional types in cstors

NinjaLikesCheez committed:
Convert tabs to spaces :///

NinjaLikesCheez committed:
Add support for borrowing/consuming keywords

NinjaLikesCheez committed:
Add support for init_declaration

NinjaLikesCheez committed:
Add support for distributed keyword/modifier

NinjaLikesCheez committed:
WIP: Add support for value and type parameter pack expr

NinjaLikesCheez committed:
Add support for parameter packs

NinjaLikesCheez committed:
Add support for package visibility modifier

NinjaLikesCheez committed:
Clean up TODOs

NinjaLikesCheez commented:
TODOs before merging:

- [ ] Update submodule (when it's released, see https://github.com/semgrep/ocaml-tree-sitter-semgrep/pull/523)

NinjaLikesCheez committed:
Add changelog and incorporate linter changes

NinjaLikesCheez committed:
Fix a bug by reversing the list before accessing the first element

NinjaLikesCheez committed:
Remove some debugging comments

NinjaLikesCheez commented on languages/swift/generic/Parse_swift_tree_sitter.ml:
```diff
@@ -84,6 +84,13 @@ let map_trailing_comma env v =
   | Some tok -> Some ((* "," *) token env tok)
   | None -> None
 
+(* TODO: NinjaLikesCheez this is horrible and should not make it to production without a hard stare in the mirror and questioning life choices *)
+let rec dotted_ident_of_type (typ : G.type_) : G.dotted_ident =
+  match typ.t with
+  | G.TyN name -> H2.dotted_ident_of_name name
+  | G.OtherType (_, [ G.T t ]) -> dotted_ident_of_type t
+  | _ -> failwith "Other types are not yet implemented"
+
```
One thing I want to highlight as a 'this needs to be resolved before merge' is this function. 

I use it to get the 'FQN-esque' name for a `CST.constrained_type`. I do this because we have a `CST.unannotated_type` and then some optional identifiers to resolve. There's already existing parsing logic for all of it's possible values which I didn't really want to rewrite. I spoke with Cooper on Slack (https://semgrep.slack.com/archives/C01LNF69F9B/p1739198829304169) and they suggested some items for me to look into (thanks Cooper for all your help!)

nmote replied:
Yeah, if I understand this correctly, we are going to have to adjust this before merging. It looks like this is called with the output of `map_unannotated_type`. It certainly looks as though it's possible for various type kinds besides `TyN` and `OtherType` to be produced there. We can't have `failwith`s happening when we encounter those cases.

I spent a little time looking at what language construct this is for. It looks esoteric and not something that I'd expect us to extract semantic information from anyway.

It looks like this is used to produce a `dotted_ident`, but that `dotted_ident` is just embedded within a `TyN` anyway. I'd suggest refactoring to plumb through a `G.type_` instead, giving us more freedom. Then use `OtherType` possibly combined with `Raw` to produce a parse tree without risk of runtime exceptions.

That will allow us to move forward with the grammar upgrade. A later change to remove uses of `raw_tree` etc. would be welcome, and it would be easier to understand and motivate in isolation with specific examples.

nmote commented on languages/swift/generic/Parse_swift_tree_sitter.ml:
```diff
@@ -3122,7 +3351,13 @@ and map_unary_expression (env : env) (x : CST.unary_expression) : G.expr =
+        | `Exp x -> map_expression env x
+        | `Choice_async _ ->
+            (* This is not valid swift code, a hack in the grammar: see: https://github.com/alex-pinkus/tree-sitter-swift/issues/468 *)
+            failwith "Invalid Swift"
```
I've encountered cases where we actually got a supposedly impossible CST after Tree Sitter's error recovery. `failwith`s like this can turn a recoverable error into an unrecoverable one. Nowadays we tend to prefer to use the raw tree construct for cases like this, which allows at least some basic matching of the untranslated section (and more importantly, does not bring down the whole parse attempt).

https://github.com/semgrep/semgrep/blob/develop/src/ast_generic/AST_generic.ml#L853-L855

NinjaLikesCheez replied:
Addressed in 1afbaa3c

nmote reviewed (commented):
Thanks for putting this together! It overall looks fairly routine, which I like to see. As you've noted, the elephant in the room is the parameter packs stuff. I've added some suggestions inline on how we can move forward.

Once we have that sorted out, I'll go over the rest with a fine-toothed comb. But at first glance it looks reasonable, so I don't expect I will have any other major feedback. Thanks again for contributing!

nmote commented:
Please re-request review once you're ready for me to take another look, so that it goes back into my review queue.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/semgrep/semgrep?pullRequest=10836) <br/>All committers have signed the CLA.

NinjaLikesCheez commented:
Oops - looks like a bad rebase on develop may have pulled in a lot of stuff I wasn't intending...

Edit: I've fixed that but I can't pull back the review request based on the code owners - sorry about that!

NinjaLikesCheez committed:
Rework map_unary_expression to not fail on a potentially invalid Swift construct

Address discussion:
https://github.com/semgrep/semgrep/pull/10836#discussion_r1960750406

kopecs commented:
The code owners one is probably due to an update we've made in the groups. No worries.

nmote commented:
Yeah, no worries. I see a bunch of updates here but since you haven't yet re-requested review I am going to assume you are still working on this.

NinjaLikesCheez commented:
> Yeah, no worries. I see a bunch of updates here but since you haven't yet re-requested review I am going to assume you are still working on this.

Yes indeed! Still have to do the most important updates based on your comments :) Will request review when done but due to limited timing I'll have to make changes when I can.

nmote commented:
No rush!

nmote commented:
@NinjaLikesCheez Still no rush, but I wanted to check in and see if you are still planning to work on this. If not, we can close the PRs for now and you can open new ones if you decide to come back to it.

NinjaLikesCheez commented:
> @NinjaLikesCheez Still no rush, but I wanted to check in and see if you are still planning to work on this. If not, we can close the PRs for now and you can open new ones if you decide to come back to it.

Hi! I do intend to get back to this - but struggling to find the time at the moment. I'll close it out and if I ever get the time I'll draft a new one.

Thank you for your all help thus far!

nmote commented:
Sounds good! Feel free to tag me/request my review when or if you get back to this.
